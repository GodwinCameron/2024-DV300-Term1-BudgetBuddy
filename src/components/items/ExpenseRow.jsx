import React from "react";

function ExpenseRow(props) {
  const { expense } = props;

  return (
    <div data-testid='expense-row' className="card card-row p-3 mb-2 row-expense">
      {expense && (
        <>
          <p>{expense.name}</p>
          <p className="amount">
            {new Intl.NumberFormat("en-za", {
              style: "currency",
              currency: "ZAR",
            }).format(expense.cost)}
          </p>
        </>
      )}
    </div>
  );
}

export default ExpenseRow;
