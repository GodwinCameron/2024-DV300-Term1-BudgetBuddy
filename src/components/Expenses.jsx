import React from "react";
import { Form, Button } from "react-bootstrap";
import ExpenseRow from "./items/ExpenseRow";
import { dummyExpenses } from "../utils";

function Expenses(props) {
  const { expenses, setExpenses } = props;

  const handleForm = (e) => {
    e.preventDefault();
    const name = document.getElementById("title").value;
    const cost = document.getElementById("expense").value;
    if (name === "" || cost === "") {
      alert("Please fill out all fields");
      return;
    }
    setExpenses([
      ...expenses,
      {
        name: name.charAt(0).toUpperCase() + name.slice(1),
        cost: Number(cost),
      },
    ]);
    document.getElementById("title").value = "";
    document.getElementById("expense").value = "";
  };

  return (
    <div>
      <h3>Household Montly Expenses</h3>
      {/* Form */}

      <div className="form-row">
        <Form.Control
          type="text"
          id="title"
          name="title"
          placeholder="Expense Title"
          autoComplete="off"
        />
        <Form.Control
          type="number"
          id="expense"
          name="expense"
          step="0.01"
          placeholder="0.00"
          autoComplete="off"
        />
        <Button onClick={handleForm} className="add-expenses">
          Add
        </Button>
      </div>

      {/* List */}
      <div className="expense-outer hide-scroll">
        {expenses &&
          expenses.map((item, index) => (
            <ExpenseRow key={index} expense={item} />
          ))}
      </div>
    </div>
  );
}

export default Expenses;
