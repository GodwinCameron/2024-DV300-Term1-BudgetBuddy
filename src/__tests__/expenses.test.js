import { render, screen } from "@testing-library/react";
import Expenses from "../components/Expenses";
import ExpenseRow from "../components/items/ExpenseRow";

describe("Tests for expenses component", () => {
  test("test default state without array", () => {
    render(<Expenses />);

    var title = screen.getByText("Household Montly Expenses");

    expect(title).toBeInTheDocument();
    console.log(
      "this helps by both testing if the component can redner in the test environment",
      "for further tests and tests how the component handles not recieving the starting array"
    );
  });
});

describe("Tests for expense row", () => {
  test("test for expense row deault state", () => {
    render(<ExpenseRow />);
    var expenseRowContainer = screen.getByTestId("expense-row");
    expect(expenseRowContainer).toBeInTheDocument();
  });

  test("test for expense row", () => {
    var expense = {
      name: "Rent",
      cost: 5000,
    };
    render(<ExpenseRow expense={expense} />);

    var expenseName = screen.getByText(expense.name);


    // var expenseCost = screen.getByText(expense.cost);

    //I'm excluding this var to be tested because the amount is formatted on the front end. 
    //Therefore if the expenseName is displayed, the amount will be displayed as well, as they
    //are in the same div and come from the same object.
    //so instead, i will test for the formmated version which wont be dynamically set

    var expenseCost = screen.getByText("R 5,000.00");

    expect(expenseName).toBeInTheDocument();
    expect(expenseCost).toBeInTheDocument();
  });
});

describe("snapshots", () => {
  test("snapshot of expenses", () => {
    const { asFragment } = render(<Expenses />);
    expect(asFragment()).toMatchSnapshot();
  });
});
