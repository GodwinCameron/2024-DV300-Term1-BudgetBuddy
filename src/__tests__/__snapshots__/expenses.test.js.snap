// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots snapshot of expenses 1`] = `
<DocumentFragment>
  <div>
    <h3>
      Household Montly Expenses
    </h3>
    <div
      class="form-row"
    >
      <input
        autocomplete="off"
        class="form-control"
        id="title"
        name="title"
        placeholder="Expense Title"
        type="text"
      />
      <input
        autocomplete="off"
        class="form-control"
        id="expense"
        name="expense"
        placeholder="0.00"
        step="0.01"
        type="number"
      />
      <button
        class="add-expenses btn btn-primary"
        type="button"
      >
        Add
      </button>
    </div>
    <div
      class="expense-outer hide-scroll"
    />
  </div>
</DocumentFragment>
`;
